### ISSUE TITLE
DTLSv1_get_timeout() returns 1 after DTLS established when acting as DTLS server

### ISSUE THREAD

**Issue by ibc** [openssl/openssl#7998]
* Working version: **OpenSSL 1.1.0h 27 Mar 2018**
* Failing version: **OpenSSL 1.1.0j 20 Nov 2018**

The scenario is as follows:

* WebRTC server using OpenSSL acting as **DTLS server**.
* The DTLS timer is managed by the app by using `DTLSv1_get_timeout()`.
* Client (browser) initiates DTLS handshake. Eventually OpenSSL's  reports `SSL_CB_HANDSHAKE_DONE` in the `SSL_CTX_set_info_callback()` callback.
* After that, the DTLS timer expires (because it was initially set, obviously) and it runs:

```c++
void DtlsTransport::OnTimer()
{
  DTLSv1_handle_timeout(this->ssl);

  // If required, send DTLS data. Pseudo code:
  int64_t read = BIO_get_mem_data(this->sslBioToNetwork, &data);
  // Send read data to the browser.
  // [...]

  BIO_reset(this->sslBioToNetwork);

  // Set the DTLS timer again.
  struct timeval dtlsTimeout{ 0, 0 };
  int64_t ret = DTLSv1_get_timeout(this->ssl, (void*)&dtlsTimeout);

  if (ret == 0)
    return;
  // Otherwise set the timer again and so on.
   [...]
}
```

It happens that **after DTLS handshake is done**, `DTLSv1_get_timeout()` returns 1 instead of 0. The app runs the `onTimer()` code above again, OpenSSL sends "not sure which pending data" to the browser, the browser replies nothing (because DTLS was already connected!), and this happens again and again until the app decides that "DTLS timeout error happened". But the fact is that **DTLS was connected**.

The issue does not happen when OpenSSL acts as **DTLS client**, neither it happens when using a previous OpenSSL version **1.1.0h 27 Mar 2018**.

Full app logs (which I hope are descriptive enough) here:

```
RTC::WebRtcTransport::MayRunDtlsTransport() | transition from DTLS local role 'auto' to 'server' and running DTLS transport +40s
RTC::WebRtcTransport::OnDtlsConnecting() | DTLS connecting +0ms
RTC::DtlsTransport::Run() | running [role:server] +1ms
RTC::DtlsTransport::OnSslInfo() | DTLS handshake start +0ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'before SSL initialization'] +0ms
RTC::DtlsTransport::OnSslInfo() | role: server, waiting:'before SSL initialization'] +0ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'before SSL initialization'] +0ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS read client hello'] +0ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS write server hello'] +0ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS write certificate'] +0ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS write key exchange'] +2ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS write certificate request'] +0ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS write server done'] +0ms
RTC::DtlsTransport::OnSslInfo() | role: server, waiting:'SSLv3/TLS write server done'] +0ms
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | --- SendPendingOutgoingDTLSData() +40s
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | 885 bytes of DTLS data ready to sent to the peer +0ms
RTC::DtlsTransport::SetTimeout() | --- SetTimeout() +1ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() returns 1 +0ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() sets dtlsTimeout to 999ms +0ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS write server done'] +4ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS read client certificate'] +0ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS read client key exchange'] +0ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS read certificate verify'] +3ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS read change cipher spec'] +0ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS read finished'] +1ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS write change cipher spec'] +0ms
RTC::DtlsTransport::OnSslInfo() | [role:server, action:'SSLv3/TLS write finished'] +0ms
RTC::DtlsTransport::OnSslInfo() | DTLS handshake done +0ms
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | --- SendPendingOutgoingDTLSData() +7ms
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | 67 bytes of DTLS data ready to sent to the peer +0ms
RTC::DtlsTransport::SetTimeout() | --- SetTimeout() +0ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() returns 1 +1ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() sets dtlsTimeout to 999ms +0ms
RTC::DtlsTransport::SetRemoteFingerprint() | handshake already done, processing it right now +2ms
RTC::DtlsTransport::CheckRemoteFingerprint() | valid remote fingerprint +0ms
RTC::DtlsTransport::GetNegotiatedSrtpProfile() | chosen SRTP profile: SRTP_AES128_CM_SHA1_80 +1ms
RTC::WebRtcTransport::OnDtlsConnected() | DTLS connected +0ms

// So DTLS has been connected!!!
// However, when the DTLS timer expires it seems that OpenSSL has yet pending data to be sent
// to the browser. But clearly, it receives nothing back so here the problem.

RTC::DtlsTransport::OnTimer() | --- OnTimer() +993ms
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | --- SendPendingOutgoingDTLSData() +0ms
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | 67 bytes of DTLS data ready to sent to the peer +1ms
RTC::DtlsTransport::SetTimeout() | --- SetTimeout() +1ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() returns 1 +0ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() sets dtlsTimeout to 1999ms +0ms
RTC::DtlsTransport::OnTimer() | --- OnTimer() +2s
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | --- SendPendingOutgoingDTLSData() +0ms
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | 67 bytes of DTLS data ready to sent to the peer +0ms
RTC::DtlsTransport::SetTimeout() | --- SetTimeout() +0ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() returns 1 +0ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() sets dtlsTimeout to 3999ms +0ms
RTC::DtlsTransport::OnTimer() | --- OnTimer() +4s
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | --- SendPendingOutgoingDTLSData() +1ms
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | 67 bytes of DTLS data ready to sent to the peer +0ms
RTC::DtlsTransport::SetTimeout() | --- SetTimeout() +0ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() returns 1 +0ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() sets dtlsTimeout to 7999ms +0ms
RTC::DtlsTransport::OnTimer() | --- OnTimer() +8s
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | --- SendPendingOutgoingDTLSData() +0ms
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | 67 bytes of DTLS data ready to sent to the peer +0ms
RTC::DtlsTransport::SetTimeout() | --- SetTimeout() +1ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() returns 1 +0ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() sets dtlsTimeout to 15999ms +0ms
RTC::DtlsTransport::OnTimer() | --- OnTimer() +16s
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | --- SendPendingOutgoingDTLSData() +0ms
RTC::DtlsTransport::SendPendingOutgoingDtlsData() | 67 bytes of DTLS data ready to sent to the peer +0ms
RTC::DtlsTransport::SetTimeout() | --- SetTimeout() +1ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() returns 1 +0ms
RTC::DtlsTransport::SetTimeout() | --- DTLSv1_get_timeout() sets dtlsTimeout to 31999ms +0ms
RTC::DtlsTransport::SetTimeout() | DTLS timeout too high (31999ms), resetting DLTS +0ms
RTC::DtlsTransport::Reset() | resetting DTLS transport +0ms
RTC::DtlsTransport::OnSslInfo() | sending DTLS warning alert: close notify +31s
RTC::WebRtcTransport::OnDtlsFailed() | DTLS fails
```

**Comment by ibc** [comment]
Hi, any update about this issue?

**Comment by mattcaswell** [comment]
I think this is due to f7506416b. I don't know where my head was at when I implemented that commit but it just looks completely wrong and should be reverted. Unfortunately doing so causes tests to fail which needs investigation.

**Comment by ibc** [comment]
Hi @mattcaswell, such a commit says:

> The timer should continue to run until we know that the client has received our last
flight (e.g. because we receive some application data).

That must be wrong. In fact, in my DTLS usage I do **not** receive any DTLS application data once the handshake is done. In my scenario (WebRTC for audio/video) the DTLS handshake is used to exchange the SRTP encryption keys between peers, nothing else.

Anyway, even in different scenarios, the DTLS server should not assume that application data will come after the handshake. According to the non existing doc:

> `DTLSv1_get_timeout()` queries the next DTLS handshake timeout. If there is a timeout in progress, it sets `*out` to the time remaining and returns one. Otherwise, it returns zero.

So, this timer is just supposed to handle the DTLS handshake, and not any next DTLS application data.

**Comment by mattcaswell** [comment]
Like I said: it should be reverted.

**Comment by ibc** [comment]
Clear, thanks.

**Comment by mattcaswell** [comment]
Fix in #8047

**Comment by ibc** [comment]
Thanks a lot.
